asmcomp/power/emit.mlp

```diff
@@ -753,74 +620,79 @@ let emit_instr env i =
         if func = env.f.fun_name then
           `	b	{emit_label env.f.fun_tailrec_entry_point_label}\n`
         else begin
-          begin match abi with
-          | ELF32 ->
-            ()
-          | ELF64v1 ->
-            emit_tocload emit_gpr 11 (TocSym func);
-            `	ld	0, 0(11)\n`;  (* code pointer *)
-            `	mtctr	0\n`;
-            `	ld	2, 8(11)\n`   (* TOC for callee *)
-          | ELF64v2 ->
-            emit_tocload emit_gpr 12 (TocSym func); (* addr of fn must be in r12 *)
-            `	mtctr	12\n`
-          end;
+          emit_tocload emit_gpr 12 (TocSym func); (* addr of fn must be in r12 *)
+          `	mtctr	12\n`;
           if env.f.fun_contains_calls then begin
-            `	{emit_string lg}	11, {emit_int(retaddr_offset env)}(1)\n`;
+            `	ld	11, {emit_int(retaddr_offset env)}(1)\n`;
             `	mtlr	11\n`
           end;
           emit_free_frame env;
-          begin match abi with
-          | ELF32 ->
-            `	b	{emit_symbol func}\n`
-          | ELF64v1 | ELF64v2 ->
-            `	bctr\n`
-          end
+          `	bctr\n`
         end
-    | Lop(Iextcall { func; alloc; }) ->
-        if not alloc then begin
-          emit_call func;
-          emit_call_nop()
+    | Lop(Iextcall { func; alloc; stack_ofs }) ->
+        if stack_ofs > 0 then begin
+          emit_tocload emit_gpr 25 (TocSym func);
+         `	li	24, {emit_int stack_ofs}\n`;
+             (* size in bytes of stack area containing the arguments *)
+          emit_call "caml_c_call_stack_args";
+          record_frame env i.live (Dbg_other i.dbg);
+          `	nop\n`
+        end else if alloc then begin
+          emit_tocload emit_gpr 25 (TocSym func);
+          emit_call "caml_c_call";
+          record_frame env i.live (Dbg_other i.dbg);
+          `	nop\n`
         end else begin
-          match abi with
-          | ELF32 ->
-            `	addis	25, 0, {emit_upper emit_symbol func}\n`;
-            `	addi	25, 25, {emit_lower emit_symbol func}\n`;
-            emit_call "caml_c_call";
-            record_frame env i.live (Dbg_other i.dbg)
-          | ELF64v1 | ELF64v2 ->
-            emit_tocload emit_gpr 25 (TocSym func);
-            emit_call "caml_c_call";
-            record_frame env i.live (Dbg_other i.dbg);
-            `	nop\n`
+          (* Save OCaml stack pointer in a callee-save register *)
+          `	mr	28, 1\n`;
+          (* Switch to C stack *)
+          let offset = Domainstate.(idx_of_field Domain_c_stack) * 8 in
+          `	ld	1, {emit_int offset}(30)\n`;
+          emit_call func;
+          emit_call_nop();
+          (* Switch back to OCaml stack *)
+          `	mr	1, 28\n`
         end
     | Lop(Istackoffset n) ->
         `	addi	1, 1, {emit_int (-n)}\n`;
         adjust_stack_offset env n
-    | Lop(Iload { memory_chunk; addressing_mode; _ }) ->
+    | Lop(Iload { memory_chunk; addressing_mode; is_atomic }) ->
```

Comment:
Non-word-sized loads will not be atomic and the compiler ensures this. The way the case is written currently does not make this obvious. Is it worth rewriting this code so that this becomes clear in the same way as we do for the RISC-V port: https://github.com/ocaml/ocaml/blob/trunk/asmcomp/riscv/emit.mlp#L356-L377? RISC-V also ensures that accidentally emitting atomic non-word-sized access is caught with a runtime assertion.

Reply:
I believe that the code that is generated for atomic loads is correct ( = implements an SC load) for any kind of load, not just a full word load.  That's why I wrote the Iload case like this.  So, an assertion "atomic => word-sized" is not needed in this port.  It's just more capable than the other ports :-)

Reply:
Ok. Sounds reasonable.